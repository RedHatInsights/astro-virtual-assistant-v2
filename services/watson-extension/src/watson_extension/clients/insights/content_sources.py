import abc
from dataclasses import dataclass
from typing import Optional, List

import injector

from watson_extension.clients import ContentSourcesURL
from watson_extension.clients.identity import AbstractUserIdentityProvider
from watson_extension.clients.platform_request import AbstractPlatformRequest


@dataclass
class GetPopularRepositoriesResponse:
    data: Optional[List[dict]]


@dataclass
class BulkCreateRepositoriesInfo:
    suggested_name: str
    distribution_arch: str
    distribution_versions: List[str]
    gpg_key: str
    metadata_verification: bool
    snapshot: bool
    url: str


@dataclass
class RepositoriesBulkCreateResponse:
    response: Optional[str] = None


class ContentSourcesClient(abc.ABC):
    @abc.abstractmethod
    async def get_popular_repositories(self) -> GetPopularRepositoriesResponse: ...

    @abc.abstractmethod
    async def repositories_bulk_create(
        self,
        repository_info: List[BulkCreateRepositoriesInfo],
    ) -> RepositoriesBulkCreateResponse: ...


class ContentSourcesClientHttp(ContentSourcesClient):
    def __init__(
        self,
        content_sources_url: injector.Inject[ContentSourcesURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.content_sources_url = content_sources_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def get_popular_repositories(self) -> GetPopularRepositoriesResponse:
        request = "/api/content-sources/v1/popular_repositories/?offset=0&limit=20"
        response = await self.platform_request.get(
            self.content_sources_url,
            request,
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        content = await response.json()

        return GetPopularRepositoriesResponse(data=content["data"])

    async def repositories_bulk_create(
        self,
        repository_info: List[BulkCreateRepositoriesInfo],
    ) -> RepositoriesBulkCreateResponse:
        headers = {"Content-Type": "application/json"}

        request = "/api/content-sources/v1.0/repositories/bulk_create/"
        response = await self.platform_request.post(
            self.content_sources_url,
            request,
            user_identity=await self.user_identity_provider.get_user_identity(),
            json=repository_info,
            headers=headers,
        )
        status = response.status

        content = await response.json()

        errors = None
        if "errors" in content:
            errors = content["errors"]

        repositories_response = None
        if status >= 400 and any(
            "already belongs" in error["detail"] for error in errors
        ):
            repositories_response = "already_enabled"
        elif status == 201:
            repositories_response = "enabled"
        else:
            repositories_response = "errors"

        return RepositoriesBulkCreateResponse(response=repositories_response)
