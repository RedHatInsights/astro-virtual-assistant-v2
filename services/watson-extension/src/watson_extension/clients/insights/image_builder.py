import abc
from dataclasses import dataclass
from typing import Optional

import injector

from watson_extension.clients import ContentSourcesURL
from watson_extension.clients.identity import AbstractUserIdentityProvider
from watson_extension.clients.platform_request import AbstractPlatformRequest


@dataclass
class EnableCustomRepositoriesResponse:
    response: Optional[str] = None


class ImageBuilderClient(abc.ABC):
    @abc.abstractmethod
    async def enable_custom_repositories(
        self, version: str
    ) -> EnableCustomRepositoriesResponse: ...


class ImageBuilderClientHttp(ImageBuilderClient):
    def __init__(
        self,
        content_sources_url: injector.Inject[ContentSourcesURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.content_sources_url = content_sources_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def enable_custom_repositories(
        self,
        version: str,
    ) -> EnableCustomRepositoriesResponse:
        request = "/api/content-sources/v1/popular_repositories/?offset=0&limit=20"
        response = await self.platform_request.get(
            self.content_sources_url,
            request,
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        content = await response.json()

        repository = None
        for repo in content["data"]:
            if repo["suggested_name"] and repo["suggested_name"].startswith(version):
                repository = repo
                break

        headers = {"Content-Type": "application/json"}
        formatted = [
            {
                "name": repository["suggested_name"],
                "distribution_arch": repository["distribution_arch"],
                "distribution_versions": repository["distribution_versions"],
                "gpg_key": repository["gpg_key"],
                "metadata_verification": repository["metadata_verification"],
                "snapshot": False,
                "url": repository["url"],
            }
        ]

        request = "/api/content-sources/v1.0/repositories/bulk_create/"
        response = await self.platform_request.post(
            self.content_sources_url,
            request,
            user_identity=await self.user_identity_provider.get_user_identity(),
            json=formatted,
            headers=headers,
        )
        status = response.status

        content = await response.json()

        errors = None
        if "errors" in content:
            errors = content["errors"]

        repositories_response = None
        if status >= 400 and any(
            "already belongs" in error["detail"] for error in errors
        ):
            repositories_response = "already_enabled"
        elif status == 201:
            repositories_response = "enabled"
        else:
            repositories_response = "errors"

        return EnableCustomRepositoriesResponse(response=repositories_response)
