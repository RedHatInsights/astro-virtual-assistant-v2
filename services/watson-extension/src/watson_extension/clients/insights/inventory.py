import abc
import enum
from dataclasses import dataclass
from typing import Optional, List

import injector

from watson_extension.clients import RHSMURL
from watson_extension.clients.identity import AbstractUserIdentityProvider
from watson_extension.clients.platform_request import AbstractPlatformRequest


@dataclass
class ActivationKeyResponse:
    ok: bool
    message: Optional[str]


class InventoryClient(abc.ABC):
    @abc.abstractmethod
    async def create_activation_key(self, name: str): ...


class InventoryClientHttp(InventoryClient):
    def __init__(
        self,
        rhsm_url: injector.Inject[RHSMURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.rhsm_url = rhsm_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def create_activation_key(self, name: str):
        body = {"name": name, "role": "", "serviceLevel": "", "usage": ""}
        response = await self.platform_request.post(
            self.rhsm_url,
            "/api/rhsm/v2/activation_keys",
            user_identity=await self.user_identity_provider.get_user_identity(),
            json=body,
        )
        response.raise_for_status()

        response_json = await response.json()

        if not response.ok:
            return ActivationKeyResponse(
                ok=response.ok,
                message=response_json.get("error", {}).get(
                    "message", "Unknown error"
                )
            )
        return ActivationKeyResponse(ok=response.ok, message=None)
