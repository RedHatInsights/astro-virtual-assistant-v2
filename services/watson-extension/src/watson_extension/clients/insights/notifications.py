import abc

import injector
import logging
from watson_extension.clients import NotificationsGWURL
from watson_extension.clients.identity import AbstractUserIdentityProvider
from watson_extension.clients.platform_request import AbstractPlatformRequest


logger = logging.getLogger(__name__)


class NotificationsClient(abc.ABC):
    @abc.abstractmethod
    async def send_notification(self, event: dict): ...


class NotificationsClientHttp(NotificationsClient):
    def __init__(
        self,
        notifications_gw_url: injector.Inject[NotificationsGWURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.notifications_gw_url = notifications_gw_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def send_notification(
        self,
        event: dict,
    ):
        headers = {"Content-Type": "application/json"}

        response = await self.platform_request.post(
            self.notifications_gw_url,
            "/notifications",
            user_identity=await self.user_identity_provider.get_user_identity(),
            json=event,
            headers=headers,
        )

        response.raise_for_status()

        return ""


class NotificationClientNoOp(NotificationsClient):
    async def send_notification(self, event: dict):
        logger.info(event)
        return ""
