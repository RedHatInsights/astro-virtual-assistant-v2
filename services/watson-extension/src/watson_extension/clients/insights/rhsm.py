import abc
from dataclasses import dataclass
from typing import Optional, List

import injector

from watson_extension.clients import RhsmURL
from watson_extension.clients.identity import AbstractUserIdentityProvider
from watson_extension.clients.platform_request import AbstractPlatformRequest


@dataclass
class SubscriptionInfo:
    number: int
    category: str


class RhsmClient(abc.ABC):
    @abc.abstractmethod
    async def check_subscriptions(
        self, category: Optional[str]
    ) -> List[SubscriptionInfo]: ...


class RhsmClientHttp(RhsmClient):
    def __init__(
        self,
        rhsm_url: injector.Inject[RhsmURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.rhsm_url = rhsm_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def check_subscriptions(
        self,
        category: Optional[str],
    ) -> List[SubscriptionInfo]:
        request = "/api/rhsm/v2/products/status"
        response = await self.platform_request.get(
            self.rhsm_url,
            request,
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        content = await response.json()
        content_body = content.get("body")

        subscriptions_info = []
        if content_body:
            if category:
                category_text = category
                if category == "expiringSoon":
                    category_text = "expiring"

                subs_category_count = content_body.get(category)
                subscriptions_info.append(
                    SubscriptionInfo(number=subs_category_count, category=category_text)
                )
            else:
                subscriptions_info = []
                if content_body.get("active"):
                    subscriptions_info.append(
                        SubscriptionInfo(
                            number=content_body["active"], category="active"
                        )
                    )
                if content_body.get("expiringSoon"):
                    subscriptions_info.append(
                        SubscriptionInfo(
                            number=content_body["expiringSoon"], category="expiring"
                        )
                    )
                if content_body.get("expired"):
                    subscriptions_info.append(
                        SubscriptionInfo(
                            number=content_body["expired"], category="expired"
                        )
                    )

        return subscriptions_info
