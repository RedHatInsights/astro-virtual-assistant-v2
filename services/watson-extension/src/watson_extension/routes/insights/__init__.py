from quart import Blueprint
from . import advisor, inventory, vulnerability, content_sources, rhsm

blueprint = Blueprint("insights", __name__, url_prefix="/insights")

blueprint.register_blueprint(advisor.blueprint)
blueprint.register_blueprint(inventory.blueprint)
blueprint.register_blueprint(vulnerability.blueprint)
blueprint.register_blueprint(content_sources.blueprint)
blueprint.register_blueprint(rhsm.blueprint)
