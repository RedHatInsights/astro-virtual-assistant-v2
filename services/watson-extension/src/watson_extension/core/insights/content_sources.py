import injector

from watson_extension.clients.insights.content_sources import ContentSourcesClient


class ContentSourcesCore:
    def __init__(self, content_sources_client: injector.Inject[ContentSourcesClient]):
        self.content_sources_client = content_sources_client

    async def enable_custom_repositories(self, version: str):
        popular_repositories = (
            await self.content_sources_client.get_popular_repositories()
        )

        repository = None
        for repo in popular_repositories.data:
            if repo["suggested_name"] and repo["suggested_name"].startswith(version):
                repository = repo
                break

        formatted_repository_info = [
            {
                "name": repository["suggested_name"],
                "distribution_arch": repository["distribution_arch"],
                "distribution_versions": repository["distribution_versions"],
                "gpg_key": repository["gpg_key"],
                "metadata_verification": repository["metadata_verification"],
                "snapshot": False,
                "url": repository["url"],
            }
        ]

        return await self.content_sources_client.repositories_bulk_create(
            formatted_repository_info
        )
