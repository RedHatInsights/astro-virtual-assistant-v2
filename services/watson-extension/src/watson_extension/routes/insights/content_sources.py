import injector
from pydantic import BaseModel

from quart import Blueprint, render_template
from quart_schema import validate_response, validate_querystring, document_headers

from watson_extension.core.insights.content_sources import (
    ContentSourcesCore,
)
from watson_extension.routes import RHSessionIdHeader

blueprint = Blueprint("content_sources", __name__, url_prefix="/content_sources")


class CustomRepositoriesRequestQuery(BaseModel):
    version: str


class EnableCustomRepositoriesResponse(BaseModel):
    response: str


@blueprint.get("/enable_custom_repositories")
@validate_querystring(CustomRepositoriesRequestQuery)
@validate_response(EnableCustomRepositoriesResponse)
@document_headers(RHSessionIdHeader)
async def enable_custom_repositories(
    query_args: CustomRepositoriesRequestQuery,
    content_sources_service: injector.Inject[ContentSourcesCore],
) -> EnableCustomRepositoriesResponse:
    version = query_args.version
    custom_repositories_response = (
        await content_sources_service.enable_custom_repositories(version)
    )

    return EnableCustomRepositoriesResponse(
        response=await render_template(
            "insights/content_sources/custom_repositories.txt.jinja",
            version=version,
            response_type=custom_repositories_response.response,
            dashboard_link="/insights/content",
        )
    )
