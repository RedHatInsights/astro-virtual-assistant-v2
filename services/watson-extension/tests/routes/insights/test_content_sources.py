from unittest.mock import MagicMock

import injector
import pytest
from quart.typing import TestClientProtocol

from watson_extension.clients.insights.content_sources import (
    GetPopularRepositoriesResponse,
    RepositoriesBulkCreateResponse,
    ContentSourcesClient,
)
from ..common import app_with_blueprint

from watson_extension.routes.insights.content_sources import blueprint
from ... import async_value


@pytest.fixture
async def content_sources_client() -> MagicMock:
    return MagicMock(ContentSourcesClient)


@pytest.fixture
async def test_client(content_sources_client) -> TestClientProtocol:
    def injector_binder(binder: injector.Binder):
        binder.bind(ContentSourcesClient, content_sources_client)

    return app_with_blueprint(blueprint, injector_binder).test_client()


async def test_custom_repositories_enabled(
    test_client, content_sources_client, snapshot
) -> None:
    content_sources_client.get_popular_repositories = MagicMock(
        return_value=async_value(
            GetPopularRepositoriesResponse(
                data=[
                    {
                        "suggested_name": "EPEL 8 repo",
                        "distribution_arch": "blah",
                        "distribution_versions": ["1"],
                        "gpg_key": "keykeykey",
                        "metadata_verification": True,
                        "url": "example.com",
                    }
                ]
            )
        )
    )

    content_sources_client.repositories_bulk_create = MagicMock(
        return_value=async_value(
            RepositoriesBulkCreateResponse(
                response="enabled",
            )
        )
    )

    response = await test_client.get(
        "/content_sources/enable_custom_repositories?version=EPEL 8",
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == snapshot


async def test_custom_repositories_already_enabled(
    test_client, content_sources_client, snapshot
) -> None:
    content_sources_client.get_popular_repositories = MagicMock(
        return_value=async_value(
            GetPopularRepositoriesResponse(
                data=[
                    {
                        "suggested_name": "EPEL 9 repo",
                        "distribution_arch": "blah",
                        "distribution_versions": ["1"],
                        "gpg_key": "keykeykey",
                        "metadata_verification": True,
                        "url": "example.com",
                    }
                ]
            )
        )
    )

    content_sources_client.repositories_bulk_create = MagicMock(
        return_value=async_value(
            RepositoriesBulkCreateResponse(
                response="already_enabled",
            )
        )
    )

    response = await test_client.get(
        "/content_sources/enable_custom_repositories?version=EPEL 9",
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == snapshot


async def test_custom_repositories_error(
    test_client, content_sources_client, snapshot
) -> None:
    content_sources_client.get_popular_repositories = MagicMock(
        return_value=async_value(
            GetPopularRepositoriesResponse(
                data=[
                    {
                        "suggested_name": "EPEL 8 repo",
                        "distribution_arch": "blah",
                        "distribution_versions": ["1"],
                        "gpg_key": "keykeykey",
                        "metadata_verification": True,
                        "url": "example.com",
                    }
                ]
            )
        )
    )

    content_sources_client.repositories_bulk_create = MagicMock(
        return_value=async_value(RepositoriesBulkCreateResponse(response=None))
    )

    response = await test_client.get(
        "/content_sources/enable_custom_repositories?version=EPEL 8",
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == snapshot
