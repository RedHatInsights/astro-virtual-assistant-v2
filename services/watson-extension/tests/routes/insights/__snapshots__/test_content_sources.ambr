# serializer version: 1
# name: test_custom_repositories_already_enabled
  'Looks like you already have the EPEL 9 repository enabled.'
# ---
# name: test_custom_repositories_enabled
  'The EPEL 8 repository has been enabled and can now be used when building images with Image Builder.'
# ---
# name: test_custom_repositories_error
  'I am sorry, I am having trouble viewing your repositories. Try visiting your repositories [here](/insights/content).'
# ---
